File: src/main/java/be/sch3lp/j2k/nullability/java/Controller.java
package be.sch3lp.j2k.nullability.java;

import static be.sch3lp.j2k.nullability.java.ISBN.isbn;

public class Controller {

    private final Library library;

    // Assume some injection framework injects a non null instance
    public Controller(final Library library) {
        this.library = library;
    }

    public BookJson lend(final String isbn) {
        return mapToJson(library.lend(isbn(isbn)));
    }

    private BookJson mapToJson(Book book) {
        return new BookJson(
                book.getAuthor(),
                book.getTitle(),
                book.getSubtitle().orElse(null),
                book.getIsbn().asString()
        );
    }

    public static class BookJson {

        private final String author;
        private final String title;
        private final String subtitle;
        private final String isbn;

        public BookJson(final String author,
                        final String title,
                        final String subtitle,
                        final String isbn) {
            this.author = author;
            this.title = title;
            this.subtitle = subtitle;
            this.isbn = isbn;
        }

        public String getAuthor() {
            return author;
        }

        public String getTitle() {
            return title;
        }

        public String getSubtitle() {
            return subtitle;
        }

        public String getIsbn() {
            return isbn;
        }
    }
}

File: src/main/java/be/sch3lp/j2k/nullability/kotlin/Controller.kt
package be.sch3lp.j2k.nullability.kotlin

class Controller(private val library: Library) {
// uncomment this to see what the Java to Kotlin code would look like
// and check which mistakes it brings to light, thanks to the compiler
//    fun lend(isbn: String?): BookJson {
//        return mapToJson(library.lend(ISBN(isbn)))
//    }
//
//    private fun mapToJson(book: Book?): BookJson {
//        return BookJson(
//            book.author,
//            book.title,
//            book.subtitle,
//            book.isbn.asString()
//        )
//    }

    fun lend(isbn: String?): BookJson? {
        return if (isbn == null) {
            null
        } else {
            library.lend(ISBN(isbn))?.asJson()
        }
    }

    private fun Book.asJson() = BookJson(
        author,
        title,
        subtitle,
        isbn.asString()
    )

    data class BookJson(
        val author: String,
        val title: String,
        val subtitle: String?,
        val isbn: String
    )
}